### PR TITLE
Setup datastore interface

### PR DESCRIPTION
**Pull request by kelvin-ngure**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes https://github.com/opensrp/fhircore/issues/2921

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [x] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/DataStoresRepository.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import timber.log.Timber
+import java.io.IOException
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.KClass
+
+@Singleton
+class DataStoresRepository @Inject constructor( private val context: Context ) {
+    companion object Keys {
+        val APP_ID = stringPreferencesKey("appId")
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/DataStoresRepository.kt:26]: Please move the strings to constants

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        get() = SerializablePractitionerDetails()
+
+    override suspend fun readFrom(input: InputStream): SerializablePractitionerDetails {
+        return try {
+            Json.decodeFromString(
+                deserializer = SerializablePractitionerDetails.serializer(),
+                string = input.readBytes().decodeToString()
+            )
+        } catch (e: SerializationException) {
+            e.printStackTrace()
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt:21]: Log using timber

Comment (dubdabasoduba) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt:21]: Can when get other exceptions while decoding the JSON?  Please catch those too

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/UserInfoDataStoreSerializer.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        get() = SerializableUserInfo()
+
+    override suspend fun readFrom(input: InputStream): SerializableUserInfo {
+        return try {
+            Json.decodeFromString(
+                    deserializer = SerializableUserInfo.serializer(),
+                    string = input.readBytes().decodeToString()
+            )
+        } catch (e: SerializationException) {
+            e.printStackTrace()
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/UserInfoDataStoreSerializer.kt:21]: Log using timber

**Review by dubdabasoduba — DISMISSED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt (hunk 2 above)
Comment (kelvin-ngure) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt:21]: > Can when get other exceptions while decoding the JSON? Please catch those too

No. Only serialization exception
image

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/UserInfoDataStoreSerializer.kt (hunk 3 above)
Comment (kelvin-ngure) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/UserInfoDataStoreSerializer.kt:21]: Done

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt (hunk 2 above)
Comment (kelvin-ngure) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/serializers/PractitionerDetailsDataStoreSerializer.kt:21]: > Log using timber

done

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/DataStoresRepository.kt (hunk 1 above)
Comment (kelvin-ngure) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/DataStoresRepository.kt]: done

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializablePractitionerDetails.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.datastore.mockdata
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class SerializablePractitionerDetails(
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializablePractitionerDetails.kt:22]: This should be migrated to the `shared.models` directory in the next phase of code refactoring. Could you rename them accordingly?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializableUserInfo.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+package org.smartregister.fhircore.engine.datastore.mockdata
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class SerializableUserInfo(
+  val name: String = "sample name",
+)
```
Comment (ellykits) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializableUserInfo.kt:24]: To be refactored in the `shared.model` directory in the engine module. Could you rename it accordingly e.g. `UserInfo`?

File: android/engine/src/test/java/org/smartregister/fhircore/engine/datastore/ProtoDataStoreTest.kt [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      }
+    }
+  }
+
+  @Test
+  fun testWritePractitionerDetails() {
+    val valueToWrite = SerializablePractitionerDetails(name = "Kelvin", id = 1)
+    runTest {
+      protoDataStore.writePractitioner(valueToWrite)
+      protoDataStore.practitioner.map { assert(it == (valueToWrite)) }
```
Comment (ellykits) [inline android/engine/src/test/java/org/smartregister/fhircore/engine/datastore/ProtoDataStoreTest.kt:64]: Do data classes compare class properties? :thinking: Please confirm and revert in the comments. I think they do (as in 2 data classes with the same content in properties should be treated as equal) so this should be okay.

File: docs/engineering/android-app/datastore/datastore.mdx [hunk 7]
```diff
@@ -0,0 +1,21 @@
+**STORING PREFERENCES DATA**
+
+Fhircore uses Preferences [DataStore](https://developer.android.com/topic/libraries/architecture/datastore) and Proto DataStore to store preferences data. Primitive types are stored in the Preferences DataStore and structured types are stored in the Proto DataStore.
+
+The interfaces exposing access to these two storage options are located in ```engine/datastore/PreferenceDataStore``` and ```engine/datastore/ProtoDataStore```.
+
+***PreferenceDataStore.kt***<br/>
+To register a key-value pair that you intend to store in the preferences datastore, add a key that enforces the type of data being stored and the name of the key in the Keys companion object.
+
+The read() method returns a flow. To observe the flow, call the function in the view model and store the resulting flow. You may then convert it to a StateFlow and expose it to the UI.
```
Comment (ellykits) [inline docs/engineering/android-app/datastore/datastore.mdx:10]:
```suggestion
The `read()` method returns a flow. To observe the flow, call the function in the view model and store the resulting flow. You may then convert it to a `StateFlow` and expose it to the UI.
```

File: docs/engineering/android-app/datastore/datastore.mdx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+The write() method can be used in the UI directly but it is preferable to expose it through a view model to follow the MVVM pattern.
```
Comment (ellykits) [inline docs/engineering/android-app/datastore/datastore.mdx:12]:
```suggestion
The `write()` method can be used in the UI directly but it is preferable to expose it through a view model to follow the MVVM pattern.
```

File: docs/engineering/android-app/datastore/datastore.mdx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+Both read() and write() are Generic methods.
```
Comment (ellykits) [inline docs/engineering/android-app/datastore/datastore.mdx:14]:
```suggestion
Both `read()` and `write()` are Generic methods.
```

File: docs/engineering/android-app/datastore/datastore.mdx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+***ProtoDataStore.kt***<br/>
+Proto DataStore allows us to store objects in a type-safe way. For ease of adaptation, we use [Kotlinx serialization](https://kotlinlang.org/docs/serialization.html#serialize-and-deserialize-json) to allow us to use Kotlin data classes instead of Protobuf files to define the schema of the data being stored.
```
Comment (ellykits) [inline docs/engineering/android-app/datastore/datastore.mdx:17]: Recommending `Kotlin` serialization (instead of `Protobuf`) is much appreciated :+1:

**Review by ellykits — CHANGES_REQUESTED** [review]
This PR looks great :+1: . Good job @kelvin-ngure and @DebbieArita. Can we provide sample code usages of the DataStore in the docs?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializableUserInfo.kt (hunk 5 above)
Comment (kelvin-ngure) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializableUserInfo.kt:24]: done

File: android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializablePractitionerDetails.kt (hunk 4 above)
Comment (kelvin-ngure) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/datastore/mockdata/SerializablePractitionerDetails.kt:22]: done

File: docs/engineering/android-app/datastore/datastore.mdx (hunk 10 above)
Comment (kelvin-ngure) [reply docs/engineering/android-app/datastore/datastore.mdx:17]: thank you!

File: android/engine/src/test/java/org/smartregister/fhircore/engine/datastore/ProtoDataStoreTest.kt (hunk 6 above)
Comment (kelvin-ngure) [reply android/engine/src/test/java/org/smartregister/fhircore/engine/datastore/ProtoDataStoreTest.kt:64]: > Do data classes compare class properties? 🤔 Please confirm and revert in the comments. I think they do (as in 2 data classes with the same content in properties should be treated as equal) so this should be okay.

yes they do. The .equals() member function is generated at compile time unless we write one ourselves in the data class

**Review by ellykits — APPROVED** [review]

**Pull request merged**